File: Tyche.Android/user/src/main/java/com/felipearpa/user/type/Username.kt
package com.felipearpa.user.type

// * The username may only contain letters, numbers, underscores, and hyphens ([a-zA-Z0-9_-]).
// * The username must be between 3 and 16 characters in length ({3,16}).
private const val usernameRegularExpression = "^[a-zA-Z0-9_-]{3,16}\$"

@JvmInline
value class Username(val value: String) {

    init {
        checkEmpty(value)
        checkUsernamePattern(value)
    }

    override fun toString(): String = value

    companion object {

        fun hasPattern(value: String): Boolean {
            return try {
                checkEmpty(value)
                checkUsernamePattern(value)
                true
            } catch (_: IllegalArgumentException) {
                false
            }
        }

        private fun checkEmpty(value: String) {
            if (value.isEmpty()) {
                throw IllegalArgumentException("Value must be not empty")
            }
        }

        private fun checkUsernamePattern(value: String) {
            val pattern = Regex(usernameRegularExpression)
            if (!pattern.containsMatchIn(value)) {
                throw IllegalArgumentException("Value must be a valid username")
            }
        }

    }

}

File: Tyche.iOS/Session/Sources/Session/Type/Username.swift
import Foundation

// * The username may only contain letters, numbers, underscores, and hyphens ([a-zA-Z0-9_-]).
// * The username must be between 3 and 16 characters in length ({3,16}).
private let usernameRegularExpression = "^[a-zA-Z0-9_-]{3,16}$"

enum UsernameError: Error {
    case empty
    case notPattern
}

public struct Username {
    let value: String
    
    public init?(_ value: String) {
        if !Username.isValid(value: value) {
            return nil
        }
        self.value = value
    }
    
    public static func isValid(value: String) -> Bool {
        do {
            try checkEmpty(value: value)
            try checkUsernamePattern(value: value)
            return true
        } catch {
            return false
        }
    }
    
    private static func checkEmpty(value: String) throws {
        if value.isEmpty {
            throw UsernameError.empty
        }
    }
    
    private static func checkUsernamePattern(value: String) throws {
        let pattern = try! NSRegularExpression(pattern: usernameRegularExpression)
        let range = NSRange(location: 0, length: value.utf16.count)
        if pattern.firstMatch(in: value, options: [], range: range) == nil {
            throw UsernameError.notPattern
        }
    }
}
